### PR TITLE
Hawkes cumulants - Tensorflow v1

### PR DESCRIPTION
**Pull request by claudio-ICL**
This PR proposes to break the class `HawkesCumulantMatching` in a base class with the general interface, and a derived class that implements the methods of the interface using tensorflow. A second derived class that will implement the same methods with pytorch is also prototyped. 
Moreover, this PR tries to migrate `HawkesCumulantMatching` to tensorflow v2 minimising the required changes: I used `tf.compat.v1`. Tests of the inference methodology are re-activated (no longer skipped by default).

### CODE REVIEW

**Review by Mbompr — APPROVED** [review]

**Comment by claudio-ICL** [comment]
Hi @Mbompr 

I am trying to adapt the code of `HawkesCumulantMatching` to tensorflow v2. There are some tests that fail however, because the numerical precision is not achieved. I was wondering whether you can explain how the expected values were produced?

**Comment by Mbompr** [comment]
If I remember correctly, these values have been obtained by making the algorithm run.
I think that if the new values are close enough, we can simply replace them. Maybe the optimization loop has changed with TF2.

**Comment by claudio-ICL** [comment]
> If I remember correctly, these values have been obtained by making the algorithm run. I think that if the new values are close enough, we can simply replace them. Maybe the optimization loop has changed with TF2.

Understood. So, they are not ground truth. From a first run with v2 I get somewhat similar values, but the precision of 6 decimal places is certainly lost.

image (1)

**Comment by Mbompr** [comment]
I think it makes sense to change the values in the test within the PR then

**Comment by claudio-ICL** [comment]
> > If I remember correctly, these values have been obtained by making the algorithm run. I think that if the new values are close enough, we can simply replace them. Maybe the optimization loop has changed with TF2.
> 
> Understood. So, they are not ground truth. From a first run with v2 I get somewhat similar values, but the precision of 6 decimal places is certainly lost.

I am thinking whether it would be more appropriate to design a test as follows: 
1. Instantiate a parametric class (e.g. exponential) with  given coefficients, and compute exact L1 norms of kernels.
2.  Generate timestamps from simulation.
3. Calibrate `HawkesCumulantMatching` on the simulated timestamps.
4. Compare calibrated `adjacency` with exact L1 norms.

**Comment by Mbompr** [comment]
I agree that makes more sense. But I find it hard to suggest something big when I had been lazy before 😬

File: tick/hawkes/inference/tests/hawkes_cumulant_matching_test.py [hunk 1]
```diff
@@ -8,212 +8,305 @@
...
+from tick.hawkes.inference.hawkes_cumulant_matching import HawkesTheoreticalCumulant
+
+
+class Test(InferenceTest):
+    def setUp(self):
+        self.dim = 2
+        np.random.seed(320982)
+
+    @staticmethod
+    def get_simulated_model():
```
Comment (claudio-ICL) [inline tick/hawkes/inference/tests/hawkes_cumulant_matching_test.py:26]: @Mbompr This is the idea that I was suggesting in the conversation of this PR. I simulate a hawkes process with exponential kernels and I train `HawkesCumulantMatching` on five samples. Loading pickled data would be replaced by this, if you agree.
More importantly, the assertion of almost equality now are between the exact theoretical values from the simulated hawkes process and the calibrated coefficients of `HawkesCumulantMatching`. When these assertion succeed, they should guarantee approximate exactness of algorithm.

File: tick/hawkes/inference/tests/hawkes_cumulant_matching_test.py [hunk 2]
```diff
@@ -8,212 +8,305 @@
...
+        if verbose:
+            print('\n')
+            print(f'expected_R_pred:\n{expected_R_pred}')
+            print(f'solution:\n{learner.solution}')
+
+        self.assertTrue(
+            np.allclose(
+                learner.solution,
+                expected_R_pred,
+                atol=0.1,  # TODO: explain why estimation is not so accurate
```
Comment (claudio-ICL) [inline tick/hawkes/inference/tests/hawkes_cumulant_matching_test.py:267]: @Mbompr Unfortunately, accuracy is not so good here. Are you familiar with the tuning of hyperparameters of `HawkesCumulantMatching`? If yes, can you please see if a better choice of `step`, `max_iter` etc... gives more accurate estimations? I can see from the paper that the accuracy was really good.

File: lib/include/tick/hawkes/inference/hawkes_cumulant.h [hunk 3]
```diff
@@ -35,4 +33,29 @@ class DLL_PUBLIC HawkesCumulant : public ModelHawkesList {
   }
 };
 
+class DLL_PUBLIC HawkesTheoreticalCumulant {
```
Comment (claudio-ICL) [inline lib/include/tick/hawkes/inference/hawkes_cumulant.h:36]: @Mbompr @PhilipDeegan 

This is a new c++ class that implements the theoretical formulae for mean intensity, covariance and skewness of hawkes processes. These are the formulae (7), (8), and (9) in the paper. 
Currently, the class is only used to test the accuracy of the estimation of `HawkesCumulantMatching`. However, the formulae have intrinsic values beside such testing,  and we might include these calculations in a more general Hawkes class in the future.

File: tick/hawkes/inference/tests/hawkes_cumulant_matching_test.py (hunk 2 above)
Comment (claudio-ICL) [reply tick/hawkes/inference/tests/hawkes_cumulant_matching_test.py:276]: Hi @Mbompr 

Could you look into this please?

**Comment by claudio-ICL** [comment]
Hi @PhilipDeegan , Do you think we can merge this?

File: lib/cpp/hawkes/inference/hawkes_cumulant.cpp [hunk 4]
```diff
@@ -126,3 +124,66 @@ double HawkesCumulant::compute_E_ijk(ulong r, ulong i, ulong j, ulong k,
   res /= (*end_times)[r];
   return res;
 }
+
+HawkesTheoreticalCumulant::HawkesTheoreticalCumulant(int dim) : d(dim) {
+  Lambda = SArrayDouble::new_ptr(dim);
```
Comment (PhilipDeegan) [inline lib/cpp/hawkes/inference/hawkes_cumulant.cpp:129]: can these be part of the constuctor setup arguments like `: d(dim)` ?

Comment (PhilipDeegan) [inline lib/cpp/hawkes/inference/hawkes_cumulant.cpp:129]: 1.  is `Lambda` the best name?
2.  variables names don't typically start with upper case

File: lib/cpp/hawkes/simulation/simu_hawkes.cpp [hunk 5]
```diff
@@ -56,15 +53,13 @@ bool Hawkes::update_time_shift_(double delay, ArrayDouble &intensity,
 void Hawkes::reset() {
   for (unsigned int i = 0; i < n_nodes; i++) {
```
Comment (PhilipDeegan) [inline lib/cpp/hawkes/simulation/simu_hawkes.cpp:54]: I know it was `unsigned int` before, I would generally prefer `std::uint32_t` or similar

File: lib/cpp/hawkes/simulation/simu_hawkes.cpp [hunk 6]
```diff
@@ -56,15 +53,13 @@ bool Hawkes::update_time_shift_(double delay, ArrayDouble &intensity,
...
     for (unsigned int j = 0; j < n_nodes; j++) {
-      if (kernels[i * n_nodes + j] != nullptr)
-        kernels[i * n_nodes + j]->rewind();
+      if (kernels[i * n_nodes + j] != nullptr) kernels[i * n_nodes + j]->rewind();
```
Comment (PhilipDeegan) [inline lib/cpp/hawkes/simulation/simu_hawkes.cpp:56]: I know it was `unsigned int` before, I would generally prefer `std::uint32_t` or similar

File: lib/include/tick/hawkes/inference/hawkes_cumulant.h [hunk 7]
```diff
@@ -35,4 +33,29 @@ class DLL_PUBLIC HawkesCumulant : public ModelHawkesList {
...
+  SArrayDoublePtr mu;
+  SArrayDoublePtr Lambda;
+  SArrayDouble2dPtr C;
+  SArrayDouble2dPtr Kc;
+  SArrayDouble2dPtr R;
+
+ public:
+  HawkesTheoreticalCumulant(int);
+  int get_dimension() { return d; }
+  void set_baseline(const SArrayDoublePtr mu) { this->mu = mu; }
```
Comment (PhilipDeegan) [inline lib/include/tick/hawkes/inference/hawkes_cumulant.h:48]: We might not have it everywhere, but `T const& n` is the preferred order

File: lib/cpp/hawkes/inference/hawkes_cumulant.cpp (hunk 4 above)
Comment (claudio-tw) [reply lib/cpp/hawkes/inference/hawkes_cumulant.cpp:129]: The variables `Lambda`, `C`, `Kc`, and `R` represent first integrated cumulant, second integrated cumulant, (reduced) third integrated cumulant, and inverse of $I - G$ (where $G$ is the matrix of $L^1$-norms of kernel functions), respectively.  Their names are directly taken from the math symbols used in the [paper](https://arxiv.org/abs/1607.06333): $\Lambda$, $C$, $K$, and $R$, respectively. 

I agree that the upper case is uncommon. I can change the name as follows:
`Lambda` -> `first_cumulant`
`C` -> `second_cumulant`
`Kc` -> `third_cumulant`
`R` -> `g_geometric`

Comment (claudio-tw) [reply lib/cpp/hawkes/inference/hawkes_cumulant.cpp]: Done here: 11353f445b59009527ca12464a5eab988bc2b844

File: lib/cpp/hawkes/simulation/simu_hawkes.cpp (hunk 6 above)
Comment (claudio-tw) [reply lib/cpp/hawkes/simulation/simu_hawkes.cpp:56]: Understood...
I will open a new PR about that because it is not related to `HawkesCumulant` or tensorflow

File: lib/include/tick/hawkes/inference/hawkes_cumulant.h (hunk 7 above)
Comment (claudio-tw) [reply lib/include/tick/hawkes/inference/hawkes_cumulant.h]: Changed: fa43f6dc9dc65013c609508c0f4fe43290fe60ac

**Comment by claudio-ICL** [comment]
@PhilipDeegan 
I believe that this PR is ready to merge. Would it be possible to do it today?
